examples/rl/ppo_cartpole.py

```diff
@@ -0,0 +1,283 @@
+value = tf.squeeze(mlp(observation_input, list(hidden_sizes) + [1], tf.tanh, None), axis=1)
+critic = keras.Model(inputs=observation_input, outputs=value)
+
+# Initialize the policy and the value function optimizers
+policy_optimizer = keras.optimizers.Adam(learning_rate=policy_learning_rate)
+value_optimizer = keras.optimizers.Adam(learning_rate=value_function_learning_rate)
+
+# Initialize the observation, episode return and episode length
+observation, episode_return, episode_length = env.reset(), 0, 0
+
+"""
+## Train
+"""
+# Iterate over the number of epochs
+for epoch in range(epochs):
+    # Initialize the sum of the returns, lengths and number of episodes for each epoch
+    sum_return = 0
+    sum_length = 0
+    num_episodes = 0
+
+    # Iterate over the steps of each epoch
+    for t in range(steps_per_epoch):
+        if render:
+            env.render()
+
+        # Get the logits, action, and take one step in the environment
+        observation = observation.reshape(1, -1)
+        logits = actor(observation)
+        action = tf.squeeze(tf.random.categorical(logits, 1), axis=1)
+        observation_new, reward, done, _ = env.step(action[0].numpy())
+        episode_return += reward
+        episode_length += 1
+
+        # Get the value and log-probability of the action
+        value_t = critic(observation)
+        logprobability_t = logprobabilities(logits, action)
+
+        # Store obs, act, rew, v_t, logp_pi_t
+        buffer.store(observation, action, reward, value_t, logprobability_t)
+
+        # Update the observation
+        observation = observation_new
+
+        # Finish trajectory if reached to a terminal state
+        terminal = done
+        if terminal or (t == steps_per_epoch - 1):
+            last_value = 0 if done else critic(observation.reshape(1, -1))
+            buffer.finish_trajectory(last_value)
+            sum_return += episode_return
+            sum_length += episode_length
+            num_episodes += 1
+            observation, episode_return, episode_length = env.reset(), 0, 0
+
+    # Get values from the buffer
+    observation_buffer, action_buffer, advantage_buffer, return_buffer, logprobability_buffer = buffer.get()
+
+    # Update the policy and implement early stopping using KL divergence
+    for _ in range(train_policy_iterations):
+        kl = train_policy(observation_buffer, action_buffer, logprobability_buffer, advantage_buffer)
+        if kl > 1.5 * target_kl:
+            # Early Stopping
+            break
+
+    # Update the value function
+    for _ in range(train_value_iterations):
+        train_value_function(observation_buffer, return_buffer)
+
+    # Print mean return and length for each epoch
+    print("\n Epoch: " + str(epoch + 1) + ". Mean Return: " + str(sum_return / num_episodes) + ". Mean Length: " + str(sum_length / num_episodes))
+
+
+"""
+## Visualizations
+
+Before training:
+
+![Imgur](https://i.imgur.com/rKXDoMC.gif)

```

Comment:
Seems to work well enough... before training?

Reply:
Is this expected?

Reply:
It does not work well. The environment resets when the pole deviates more than 15 degrees from the vertical axis. However, if you do not know it, it's not easy to understand it. If you see the other two gifs the movement is much more smooth because the environment reset is not so frequent. So, the discontinuity in the "before training" gif shows that the random agent does not work well.

Reply:
Got it.